plugins/ngrok/credentials.go

```diff
@@ -0,0 +1,102 @@
+package ngrok
+
+import (
+	"context"
+
+	"github.com/1Password/shell-plugins/sdk"
+	"github.com/1Password/shell-plugins/sdk/importer"
+	"github.com/1Password/shell-plugins/sdk/provision"
+	"github.com/1Password/shell-plugins/sdk/schema"
+	"github.com/1Password/shell-plugins/sdk/schema/credname"
+	"github.com/1Password/shell-plugins/sdk/schema/fieldname"
+	"gopkg.in/yaml.v3"
+)
+
+func Credentials() schema.CredentialType {
+	return schema.CredentialType{
+		Name:    credname.Credentials,
+		DocsURL: sdk.URL("https://ngrok.com/docs/ngrok-agent/config"),
+		Fields: []schema.CredentialField{
+			{
+				Name:                fieldname.AuthToken,
+				MarkdownDescription: "Auth Token used to authenticate to ngrok.",
+				Secret:              true,
+				Composition: &schema.ValueComposition{
+					Length: 43,
+					Charset: schema.Charset{
+						Uppercase: true,
+						Lowercase: true,
+						Digits:    true,
+					},
+				},
+			},
+			{
+				Name:                fieldname.APIKey,
+				MarkdownDescription: "API Key used to authenticate to ngrok API.",
+				Secret:              true,
+				Composition: &schema.ValueComposition{
+					Length: 48,
+					Charset: schema.Charset{
+						Uppercase: true,
+						Lowercase: true,
+						Digits:    true,
+					},
+				},
+			},
+		},
+		DefaultProvisioner: provision.TempFile(ngrokConfig, provision.Filename("config.yml"), provision.AddArgs("--config", "{{ .Path }}")),
+		Importer: importer.TryAll(
+			importer.TryEnvVarPair(defaultEnvVarMapping),
+			importer.MacOnly(
+				TryngrokConfigFile("~/Library/Application Support/ngrok/ngrok.yml"),
+			),
+			importer.LinuxOnly(
+				TryngrokConfigFile("~/.config/ngrok/ngrok.yml"),
+			),
+		)}
+}
+
+func ngrokConfig(in sdk.ProvisionInput) ([]byte, error) {
+	config := Config{
+		AuthToken: in.ItemFields[fieldname.AuthToken],
+		APIKey:    in.ItemFields[fieldname.APIKey],
+		Version:   "2", // required field for ngrok CLI to work
+	}
+	contents, err := yaml.Marshal(&config)
+	if err != nil {
+		return nil, err
+	}
+	return []byte(contents), nil
+}
+
+var defaultEnvVarMapping = map[string]sdk.FieldName{
+	"NGROK_AUTHTOKEN": fieldname.AuthToken,
+	"NGROK_API_KEY":   fieldname.APIKey,
+}
```

Comment:
If this is a valid way to specify credentials, why not provision them as environment variables as well? This would get rid of having to somehow determine the config file version, and the workaround of needing to hardcode it in the Config passed to the process. 

Also, file provisioning comes with some design limitations related to the possibility of repeated reading, which is not a concern when using environment variables. Whenever possible, we should provision envvars.

Reply:
Thanks for the fast review, @hculea! I made a note on my PR but just to reiterate, the API Key environment variable isn’t working as expected. Seems to be an issue with ngrok CLI and I’ve asked about it here:

https://github.com/inconshreveable/ngrok/issues/895

That’s the only reason I chose file based provisioning.

Reply:
My apologies, seems like I did not read through with your MR description. Thanks for clarifying! I'm looking forward to see the response of folks from `ngrok`.

Reply:
No problem at all! I’m not sure if we’d hear back soon because some issues seem to indicate that issue tracker isn’t maintained. I’ll try asking on their slack but in the event we don’t hear soon, are we in agreement to go ahead with file provisioning approach?

Reply:
Sounds good! 

One more suggestion - I am not sure the repo where you have opened the issue is the currently maintained version of ngrok? (although it has 22.5k stars)

However, by their [changelogs](https://ngrok.com/docs/ngrok-agent/changelog#ngrok-agent-311---2023-01-13), it seems that the last released version was on 13th of January.

You might have more luck opening an issue on one of the repos within this org: https://github.com/ngrok

Hope this helps!

Reply:
I asked on one of `ngrok` organization's repository but no response:

- https://github.com/ngrok/ngrok-api-go/issues/20

I have given one last attempt [here on their Slack](https://ngrokcommunity.slack.com/archives/C033D4W1R6E/p1675671666308269).

I also had a look at `[ngrok-api-go](https://github.com/ngrok/ngrok-api-go/)` code but couldn't find any references to `NGROK_API_TOKEN` environment variable, which suggests that the [documentation here](https://ngrok.com/docs/ngrok-agent#supported-env-variables) may be outdated. I suggest we wait for one of ngrok's staff to respond to my Slack ping.

Reply:
Hi @hculea, I have some updates:

<img width="456" alt="image" src="https://user-images.githubusercontent.com/18581859/217431266-066f5858-9067-4d20-80eb-609c8e7ce76e.png">

- ngrok staff confirmed ([on their Slack](https://ngrokcommunity.slack.com/archives/C033D4W1R6E/p1675725230866699?thread_ts=1675671666.308269&cid=C033D4W1R6E)) that environment variable isn't supported for API part of the ngrok agent for now. Auth Token (which is used for the rest of the ngrok agent features) is though.
- They removed it from the docs: https://github.com/ngrok/ngrok-docs/pull/188

At this point, my suggestion is to move forward with file-based provisioning, because at least that way, API key can be used. Once they announce enviroment variable-based authentication for API as well, we can our provisioning.

What do you think? Happy to take a different approach if you feel so. 

This is the last blocker for this PR to be merged.

Reply:
I responded to this on a wider note here: https://github.com/1Password/shell-plugins/pull/165#issuecomment-1427661811